Merge novel-allele-and-genotype branch to dev (#458)
nf-core/airrflow: docs/usage/genotyping_tutorial.md

# nf-core/airrflow: novel allele detection and genotype inference tutorial

This tutorial provides an introduction on how to add novel allele detection and genotype inference to an nf-core/airrflow pipeline run. These optional steps are performed with [TIgGER](https://tigger.readthedocs.io/en/stable/) and are only applicable to BCR sequences.

## Pre-requisites

Novel allele detection and genotype inference are optional add-on steps in nf-core/airrflow pipeline. Before using these features, make sure you are familiar with the basic usage of the pipeline. If not, we recommend reviewing the bulk or single-cell tutorials before proceeding with this guide.

## How to add novel allele detection and genotype inference in the pipeline

> To include novel allele detection and genotype inference to the pipeline, just add flag `--genotyping` in your command.

Reviewer: ```suggestion
To execute the novel allele detection and genotype inference nf-core/airrflow functionality, just add flag `--genotyping` in your command.
```

> 

Reviewer: I'm missing here a short list of the steps that airrflow performs for genotyping.

## Testing novel allele detection and genotype inference with built-in tests

If you have set up Nextflow and Docker for your local machine, test nf-core/airrflow novel allele detection and genotype inference with the built-in test.

```bash

> nextflow run nf-core/airrflow -r 5.0.0 -profile test_genotyping_small,docker --outdir test_genotyping_results

Reviewer: ```suggestion
nextflow run nf-core/airrflow -r 5.1.0 -profile test_genotyping_small,docker --outdir test_genotyping_results
```

```

Change the `docker` profile to `singularity` if you use Codespaces since Docker currently cannot be used in Codespaces. You can first set up a Singularity cache directory which will allow the reuse of Singularity container across all runs:

```bash
mkdir singularity_cache
export NXF_SINGULARITY_CACHEDIR="/workspaces/airrflow/singularity_cache"
```

Then run nf-core/airrflow with the genotyping test data:

```bash

> nextflow run nf-core/airrflow -r 5.0.0 -profile test_genotyping_small,singularity --outdir test_genotyping_results

Reviewer: ```suggestion
nextflow run nf-core/airrflow -r 5.1.0 -profile test_genotyping_small,singularity --outdir test_genotyping_results
```

Plots of three evidence for novel allele IGHV1-24\*01_G9A:

<p align="center">
  <img src="../images/novel_evidence.png" style="width:70%;">
</p>

2. Genotype inference plots can be found in html report in the folder 'novel_alleles_and_genotyping/02-genotype_inference/subject_id/subject_id_bayesian_genotype_inference_report/index.html'.

## Genotyping related flags in nf-core/airrflow pipeline

1. `--genotyping`: Perform TIgGER genotype inference if it is set to be `true`. It is `false` by default.
2. `--single_clone_representative`: Keep only one sequence in a clone to represent all the sequences in the same clone if it is set to be `true`. If `--genotyping` is `true`, `single_clone_representative` is `true` by default.
3. `--genotyping_clonal_threshold`: Threshold for determining if two sequences come from the same clone or not while inferring clones to find single clone representative before genotype inference. Default value is 0.2.

> 4. `--novel_allele_inference`: whether to perform TIgGER novel allele inference. If `--genotyping` is `true`, `--single_clone_representative` is `true` by default.

Reviewer: This section here feels out of place and too detailed for this tutorial. Could you instead create a new section within the `usage.md` docs called `Important considerations for genotyping and novel allele inference`. And add some extra information there. Importantly that it only supports targeted BCR for now, and that due to the large read coverage needed probably single-cell doesn't have enough coverage for it. And then you can also specifiy some details regarding the other flags for genotyping. Basically mirroring the section we have there for already `important considerations for clonal analysis`.
